From 74577808497a5f4ce3e279d1b13bab07877b6560 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@st.com>
Date: Wed, 9 Mar 2022 18:21:36 +0100
Subject: [PATCH 0246/1695] drivers: stm32_rifsc: add execution context TDCID
 check

Adds stm32_rifsc_check_tdcid() function that checks if execution context
is TDCID or not. This information is available in RIFSC_RIMC_CR register.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8333bfc277cb8730aafa32ffea4d1e55d5a487e4 drivers: stm32_rifsc: add execution context TDCID check]
Change-Id: I5a72d98ac273f76c5a250f61c3289fb6b33a7f9c
Signed-off-by: Gatien Chevallier <gatien.chevallier@st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 48 ++++++++++++++++++++---------
 core/include/drivers/stm32_rif.h    | 17 ++++++++++
 2 files changed, 51 insertions(+), 14 deletions(-)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index edf6d54fc..7fbe736ac 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -22,6 +22,7 @@
 #define _RIFSC_RISC_SECCFGR0		U(0x10)
 #define _RIFSC_RISC_PRIVCFGR0		U(0x30)
 #define _RIFSC_RISC_PER0_CIDCFGR	U(0x100)
+#define _RIFSC_RIMC_CR			U(0xC00)
 #define _RIFSC_RIMC_ATTR0		U(0xC10)
 
 #define _RIFSC_HWCFGR3			U(0xFE8)
@@ -31,34 +32,39 @@
 
 /* RIFSC_HWCFGR2 register fields */
 #define _RIFSC_HWCFGR2_CFG1_MASK	GENMASK_32(15, 0)
-#define _RIFSC_HWCFGR2_CFG1_SHIFT	0
+#define _RIFSC_HWCFGR2_CFG1_SHIFT	U(0)
 #define _RIFSC_HWCFGR2_CFG2_MASK	GENMASK_32(23, 16)
-#define _RIFSC_HWCFGR2_CFG2_SHIFT	16
+#define _RIFSC_HWCFGR2_CFG2_SHIFT	U(16)
 #define _RIFSC_HWCFGR2_CFG3_MASK	GENMASK_32(31, 24)
-#define _RIFSC_HWCFGR2_CFG3_SHIFT	24
+#define _RIFSC_HWCFGR2_CFG3_SHIFT	U(24)
 
 /* RIFSC_HWCFGR1 register fields */
 #define _RIFSC_HWCFGR1_CFG1_MASK	GENMASK_32(3, 0)
-#define _RIFSC_HWCFGR1_CFG1_SHIFT	0
+#define _RIFSC_HWCFGR1_CFG1_SHIFT	U(0)
 #define _RIFSC_HWCFGR1_CFG2_MASK	GENMASK_32(7, 4)
-#define _RIFSC_HWCFGR1_CFG2_SHIFT	4
+#define _RIFSC_HWCFGR1_CFG2_SHIFT	U(4)
 #define _RIFSC_HWCFGR1_CFG3_MASK	GENMASK_32(11, 8)
-#define _RIFSC_HWCFGR1_CFG3_SHIFT	8
+#define _RIFSC_HWCFGR1_CFG3_SHIFT	U(8)
 #define _RIFSC_HWCFGR1_CFG4_MASK	GENMASK_32(15, 12)
-#define _RIFSC_HWCFGR1_CFG4_SHIFT	12
+#define _RIFSC_HWCFGR1_CFG4_SHIFT	U(12)
 #define _RIFSC_HWCFGR1_CFG5_MASK	GENMASK_32(19, 16)
-#define _RIFSC_HWCFGR1_CFG5_SHIFT	16
+#define _RIFSC_HWCFGR1_CFG5_SHIFT	U(16)
 #define _RIFSC_HWCFGR1_CFG6_MASK	GENMASK_32(23, 20)
-#define _RIFSC_HWCFGR1_CFG6_SHIFT	20
+#define _RIFSC_HWCFGR1_CFG6_SHIFT	U(20)
+
+/*
+ * RIMC_CR register fields
+ */
+#define _RIFSC_RIMC_CR_TDCID_MASK	GENMASK_32(6, 4)
 
 /* RIFSC_VERR register fields */
 #define _RIFSC_VERR_MINREV_MASK		GENMASK_32(3, 0)
-#define _RIFSC_VERR_MINREV_SHIFT	0
+#define _RIFSC_VERR_MINREV_SHIFT	U(0)
 #define _RIFSC_VERR_MAJREV_MASK		GENMASK_32(7, 4)
-#define _RIFSC_VERR_MAJREV_SHIFT	4
+#define _RIFSC_VERR_MAJREV_SHIFT	U(4)
 
 /* Periph id per register */
-#define _PERIPH_IDS_PER_REG		32
+#define _PERIPH_IDS_PER_REG		U(32)
 #define _OFST_PERX_CIDCFGR		U(0x8)
 
 #define RIFSC_RISC_CFEN_MASK		BIT(0)
@@ -87,8 +93,8 @@
 					 RIFSC_RIMC_MPRIV_MASK)
 
 /* max entries */
-#define MAX_RIMU	16
-#define MAX_RISUP	128
+#define MAX_RIMU			U(16)
+#define MAX_RISUP			U(128)
 
 #define _RIF_FLD_PREP(field, value)	(((uint32_t)(value) << (field ## _SHIFT)) & (field ## _MASK))
 #define _RIF_FLD_GET(field, value)	(((uint32_t)(value) & (field ## _MASK)) >> (field ## _SHIFT))
@@ -424,6 +430,20 @@ static TEE_Result stm32_rimu_setup(struct rifsc_platdata *pdata)
 	return 0;
 }
 
+TEE_Result stm32_rifsc_check_tdcid(bool *tdcid_state)
+{
+	if (!rifsc_pdata.base)
+		return TEE_ERROR_DEFER_DRIVER_INIT;
+
+	*tdcid_state = false;
+
+	if (((io_read32(rifsc_pdata.base + _RIFSC_RIMC_CR) &
+	     _RIFSC_RIMC_CR_TDCID_MASK)) == (RIF_CID1 << SCID_SHIFT))
+		*tdcid_state = true;
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result
 stm32_rifsc_get_data_config(const void *fdt, int node,
 			    int *ndata, void **data_cfg)
diff --git a/core/include/drivers/stm32_rif.h b/core/include/drivers/stm32_rif.h
index 0a073614e..a72c38254 100644
--- a/core/include/drivers/stm32_rif.h
+++ b/core/include/drivers/stm32_rif.h
@@ -79,6 +79,23 @@ struct rif_conf_data {
 	uint32_t *lock_conf;
 };
 
+/* This function checks if execution context is TDCID or not.
+ * @tdcid_state: [out] Set to true if TDCID, false otherwise.
+ * Returns TEE_ERROR_DEFER_DRIVER_INIT if RIFSC driver isn't probed, TEE_SUCCESS
+ * otherwise.
+ */
+#ifdef CFG_STM32_RIF
+TEE_Result stm32_rifsc_check_tdcid(bool *tdcid_state);
+#else
+static inline TEE_Result stm32_rifsc_check_tdcid(bool *tdcid_state)
+{
+	/* Without CFG_STM32_RIF every CPU can behave as TDCID */
+	*tdcid_state = true;
+
+	return TEE_SUCCESS;
+}
+#endif
+
 /*
  * Check every possible configuration where accessing RIF
  * is possible :
-- 
2.47.2

